File: services/vision-worker/core/events.py
import time


ALLOWED_EVENT_TYPES = {
    "vision.identity",
    "vision.unknown",
    "vision.uncertain",
    "vision.motion",
    "vision.weapon",
    "vision.fall",
    "vision.tamper",
}


class EventBuilder:

    VERSION = 1

    def __init__(self, context=None):

        self.context = dict(context or {})

    # ------------------------------------------------

    def set_context(self, **context):

        clean = {
            key: value
            for key, value in context.items()
            if value is not None and value != ""
        }

        self.context = clean

    # ------------------------------------------------

    def clear_context(self):

        self.context = {}

    # ------------------------------------------------

    def _base(self, event_type, camera, scene_id, track_id=None, payload=None):

        if event_type not in ALLOWED_EVENT_TYPES:
            raise ValueError(f"unsupported vision event type: {event_type}")

        payload = payload or {}

        camera_id = self.context.get("camera_id") or camera
        device_id = self.context.get("device_id") or camera_id

        payload.setdefault("camera_id", camera_id)
        payload.setdefault("camera", camera_id)
        payload.setdefault("device_id", device_id)
        payload.setdefault("scene_id", scene_id)

        for key in (
            "node_id",
            "clip_id",
            "clip_path",
        ):
            value = self.context.get(key)
            if value is not None and value != "":
                payload.setdefault(key, value)

        if track_id is not None:
            payload.setdefault("track_id", str(track_id))

        timestamp = time.time()
        payload.setdefault("timestamp", timestamp)

        event = {
            "type": event_type,
            "source": device_id,
            "camera_id": camera_id,
            "device_id": device_id,
            "scene_id": scene_id,
            "timestamp": timestamp,
            "version": self.VERSION,
            "payload": payload
        }

        if self.context.get("node_id"):
            event["node_id"] = self.context["node_id"]

        if self.context.get("clip_id"):
            event["clip_id"] = self.context["clip_id"]

        if track_id is not None:
            event["track_id"] = str(track_id)

        identity = payload.get("identity")
        if identity:
            event["identity"] = identity

        if "confidence" in payload:
            event["confidence"] = float(payload["confidence"])

        return event

    # ------------------------------------------------

    def _scene(self, weapon=False, weapon_confidence=0.0, fall=False):

        return {
            "weapon": bool(weapon),
            "weapon_confidence": float(weapon_confidence),
            "fall": bool(fall)
        }

    # ------------------------------------------------

    def identity(self, camera, scene_id, track_id, person, score, persons,
                 weapon=False, weapon_confidence=0.0, fall=False):

        payload = {
            "identity": person,
            "resident_id": person,
            "confidence": float(score),
            "persons": int(persons),
            "scene": self._scene(weapon, weapon_confidence, fall)
        }

        return self._base(
            "vision.identity",
            camera,
            scene_id,
            track_id,
            payload
        )

    # ------------------------------------------------

    def unknown(self, camera, scene_id, track_id, persons,
                weapon=False, weapon_confidence=0.0, fall=False):

        payload = {
            "identity": "unknown",
            "confidence": 0.0,
            "persons": int(persons),
            "scene": self._scene(weapon, weapon_confidence, fall)
        }

        return self._base(
            "vision.unknown",
            camera,
            scene_id,
            track_id,
            payload
        )

    # ------------------------------------------------

    def uncertain(self, camera, scene_id, track_id, identity, score, persons,
                  weapon=False, weapon_confidence=0.0, fall=False):

        payload = {
            "identity": "uncertain",
            "best_match": identity,
            "resident_id": identity,
            "confidence": float(score),
            "persons": int(persons),
            "scene": self._scene(weapon, weapon_confidence, fall)
        }

        return self._base(
            "vision.uncertain",
            camera,
            scene_id,
            track_id,
            payload
        )

    # ------------------------------------------------

    def motion(self, camera, scene_id):

        return self._base(
            "vision.motion",
            camera,
            scene_id,
            payload={
                "motion": True,
                "confidence": 1.0,
            },
        )

    # ------------------------------------------------

    def tamper(self, camera, scene_id):

        return self._base(
            "vision.tamper",
            camera,
            scene_id,
            payload={
                "tamper": True,
                "confidence": 1.0,
            },
        )

    # ------------------------------------------------

    def weapon(self, camera, scene_id, track_id=None, confidence=0.0, weapon_type="weapon"):

        return self._base(
            "vision.weapon",
            camera,
            scene_id,
            track_id,
            {
                "weapon": weapon_type,
                "weapon_type": weapon_type,
                "confidence": float(confidence),
            }
        )

    # ------------------------------------------------

    def weapon_firearm(self, camera, scene_id, track_id=None, confidence=0.0):

        return self.weapon(
            camera,
            scene_id,
            track_id,
            confidence,
            "firearm",
        )

    # ------------------------------------------------

    def fall(self, camera, scene_id, track_id=None, confidence=1.0):

        return self._base(
            "vision.fall",
            camera,
            scene_id,
            track_id,
            {
                "fall": True,
                "confidence": float(confidence),
            }
        )

    # ------------------------------------------------

    def fall_detected(self, camera, scene_id, track_id=None, confidence=1.0):

        return self.fall(
            camera,
            scene_id,
            track_id,
            confidence,
        )

File: internal/simulation/events.go
package simulation

import (
	"encoding/json"
	"strings"
	"time"

	"synora/pkg/contract"
)

const (
	defaultDevice     = "cam_01"
	defaultIdentity   = "alexis"
	defaultConfidence = 0.92
)

func BuildEventsForScenario(run SimulationRun, scenario Scenario, overrides EventBuildOptions) ([]contract.Message, error) {
	out := make([]contract.Message, 0, len(scenario.Steps))
	for _, step := range scenario.Steps {
		msg, err := BuildMessage(EventBuildOptions{
			Type:         step.EventType,
			Source:       overrides.Source,
			SourceType:   overrides.SourceType,
			DeviceID:     firstNonEmpty(overrides.DeviceID, step.DeviceID),
			CameraID:     firstNonEmpty(overrides.CameraID, step.CameraID),
			NodeID:       firstNonEmpty(overrides.NodeID, step.NodeID),
			Identity:     firstNonEmpty(step.Identity, overrides.Identity),
			Confidence:   nonZeroFloat(step.Confidence, overrides.Confidence),
			Run:          &run,
			ScenarioID:   scenario.ID,
			StepID:       step.ID,
			DryRun:       overrides.DryRun || run.Mode == ModeDryRun,
			GeneratedBy:  firstNonEmpty(overrides.GeneratedBy, run.CreatedBy),
			LearningMode: overrides.LearningMode,
			Data:         step.Data,
		})
		if err != nil {
			return nil, err
		}
		out = append(out, msg)
	}
	return out, nil
}

func BuildMessage(opts EventBuildOptions) (contract.Message, error) {
	eventType := contract.NormalizeEventType(opts.Type)
	payload := BuildPayload(opts)
	body, err := json.Marshal(payload)
	if err != nil {
		return contract.Message{}, err
	}
	now := opts.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}
	source := firstNonEmpty(opts.Source, transportSourceForGenerator(opts.GeneratedBy), "lab")
	sourceType := firstNonEmpty(opts.SourceType, contract.SourceSimulator)
	return contract.Message{
		Type:       eventType,
		Kind:       contract.KindEvent,
		Source:     source,
		Target:     "core",
		SourceType: sourceType,
		Timestamp:  now,
		Priority:   contract.EventPriority(eventType),
		Payload:    body,
	}, nil
}

func BuildPayload(opts EventBuildOptions) map[string]any {
	now := opts.Now
	if now.IsZero() {
		now = time.Now().UTC()
	}
	eventType := contract.NormalizeEventType(opts.Type)
	deviceID := firstNonEmpty(opts.DeviceID, opts.CameraID, defaultDevice)
	cameraID := firstNonEmpty(opts.CameraID, deviceID)
	trackID := opts.TrackID
	if trackID == "" {
		trackID = "track-" + cameraID
	}
	clipID := opts.ClipID
	if clipID == "" {
		clipID = "clip-" + cameraID
	}
	payload := cloneMap(opts.Data)
	if payload == nil {
		payload = map[string]any{}
	}
	payload["device_id"] = deviceID
	payload["camera_id"] = cameraID
	payload["node_id"] = opts.NodeID
	payload["track_id"] = trackID
	payload["clip_id"] = clipID
	payload["timestamp"] = now.Format(time.RFC3339Nano)
	if opts.ClipPath != "" {
		payload["clip_path"] = opts.ClipPath
	}
	switch eventType {
	case contract.EventVisionIdentity:
		payload["identity"] = firstNonEmpty(opts.Identity, defaultIdentity)
		payload["confidence"] = nonZeroFloat(opts.Confidence, defaultConfidence)
	case contract.EventVisionUnknown:
		payload["identity"] = "unknown"
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.70)
	case contract.EventVisionUncertain:
		payload["identity"] = "uncertain"
		if opts.Identity != "" {
			payload["best_match"] = opts.Identity
		}
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.62)
	case contract.EventVisionMotion:
		payload["motion"] = true
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.80)
	case contract.EventVisionWeapon:
		payload["weapon"] = true
		payload["weapon_type"] = "unknown"
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.88)
	case contract.EventVisionFall:
		payload["fall"] = true
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.82)
	case contract.EventVisionTamper:
		payload["tamper"] = true
		payload["confidence"] = nonZeroFloat(opts.Confidence, 0.82)
	case contract.EventDeviceOffline, contract.EventDiscoveryCameraOffline:
		payload["online"] = false
	case contract.EventDiscoveryCameraOnline:
		payload["online"] = true
	}
	payload["metadata"] = EventMetadata(opts)
	return payload
}

func EventMetadata(opts EventBuildOptions) map[string]any {
	generatedBy := firstNonEmpty(opts.GeneratedBy, GeneratedBySimulationEngine)
	testRunID := ""
	scenarioID := opts.ScenarioID
	if opts.Run != nil {
		testRunID = opts.Run.ID
		if scenarioID == "" {
			scenarioID = opts.Run.ScenarioID
		}
		if generatedBy == GeneratedBySimulationEngine && opts.Run.CreatedBy != "" {
			generatedBy = opts.Run.CreatedBy
		}
	}
	eventInstanceID := firstNonEmpty(opts.EventInstanceID, simulationEventInstanceID(testRunID, opts.StepID))
	return map[string]any{
		"simulated":         true,
		"test_run_id":       testRunID,
		"scenario_id":       scenarioID,
		"scenario_step_id":  opts.StepID,
		"event_instance_id": eventInstanceID,
		"dry_run":           opts.DryRun,
		"generated_by":      generatedBy,
		"learning_mode":     firstNonEmpty(opts.LearningMode, "simulation"),
	}
}

func simulationEventInstanceID(testRunID string, stepID string) string {
	testRunID = strings.TrimSpace(testRunID)
	stepID = strings.TrimSpace(stepID)
	switch {
	case testRunID != "" && stepID != "":
		return testRunID + ":" + stepID
	case testRunID != "":
		return testRunID
	case stepID != "":
		return stepID
	default:
		return ""
	}
}

func cloneMap(source map[string]any) map[string]any {
	if source == nil {
		return nil
	}
	out := make(map[string]any, len(source))
	for key, value := range source {
		out[key] = value
	}
	return out
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return strings.TrimSpace(value)
		}
	}
	return ""
}

func nonZeroFloat(value float64, fallback float64) float64 {
	if value == 0 {
		return fallback
	}
	return value
}

func transportSourceForGenerator(generatedBy string) string {
	switch strings.TrimSpace(generatedBy) {
	case GeneratedBySynoraLab:
		return "lab"
	case GeneratedBySynoraAPI:
		return "api"
	case GeneratedByFrontendTestMode:
		return "api"
	case GeneratedBySimulationEngine:
		return "simulation"
	default:
		return ""
	}
}
